src/misc/messages.cpp

```diff
@@ -62,47 +83,86 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp);
+			assert(rendered_cp_msg[cp].length());
+		}
+		return rendered_cp_msg[cp].c_str();
 	}
 
-	void Set(const char *markup)
+	void Set(const char *markup, const bool is_utf8)
 	{
 		assert(markup);
-		markup_msg = markup;
+		markup_msg  = markup;
+		is_utf8_msg = is_utf8;
+
 		rendered_msg.clear();
+		markup_cp_msg.clear();
+		rendered_cp_msg.clear();
 	}
 };
 
 static std::unordered_map<std::string, Message> messages;
 static std::deque<std::string> messages_order;
 
-// Add the message if it doesn't exist yet
-void MSG_Add(const char *name, const char *markup_msg)
+static void msg_add(const char *name, const char *markup_msg, const bool is_utf8)
 {
-	const auto &pair = messages.try_emplace(name, markup_msg);
+	const auto &pair = messages.try_emplace(name, markup_msg, is_utf8);
 	if (pair.second) // if the insertion was successful
 		messages_order.emplace_back(name);
 }
 
+// Add the message if it doesn't exist yet
+void MSG_Add(const char *name, const char *markup_msg)
+{
+	msg_add(name, markup_msg, false);
+}
+
 // Replace existing or add if it doesn't exist
-void MSG_Replace(const char *name, const char *markup_msg)
+static void msg_replace(const char *name, const char *markup_msg, const bool is_utf8)
 {
 	auto it = messages.find(name);
 	if (it == messages.end())
-		MSG_Add(name, markup_msg);
+		msg_add(name, markup_msg, is_utf8);
 	else
-		it->second.Set(markup_msg);
+		it->second.Set(markup_msg, is_utf8);
 }
 
-static bool LoadMessageFile(const std_fs::path &filename)
+static bool load_message_file(const std_fs::path &filename_plain,
+                              const std_fs::path &filename_utf8)
 {
-	if (filename.empty())
+	if (filename_plain.empty() && filename_utf8.empty())
 		return false;
 
-	const auto filename_str = filename.string();
+	auto check_file = [](const std_fs::path &filename) {
+		if (filename.empty())
+			return false;
+		return (std_fs::status(filename).type() !=
+		        std_fs::file_type::not_found);
+	};
+
+	bool is_utf8 = false;
+
+	std::string filename_str;
+	if (check_file(filename_utf8)) {
+		filename_str = filename_utf8.string();
+		is_utf8      = true;
+	} else if (check_file(filename_plain))
+		filename_str = filename_plain.string();
+	else {
+		if (filename_utf8.empty())
+			filename_str = filename_plain.string();
+		else if (filename_plain.empty())
+			filename_str = filename_utf8.string();
+		else
+			filename_str = filename_plain.string() + " / " +
+			               filename_utf8.string();
```

Comment:
filesystem::path's have a `/` operator that can combine any OS-specific path element (including the odds and ends, like a root drive in Windows or MorphOS) So doing something like: `c_drive_root / dir` takes care of all those Windows-ism's like the :\ drive and double-back slashes.

In this case, keep them as `path` objects until the final string conversion:

 `(filename_plain / filename_utf8).string();`

Reply:
Here this is used just for logging purposes, we are not constructing the path here. I have changed the code a little so that it is less confusing.